### DISCUSSION TITLE
Cant get into how you do ore control. Help?

### DISCUSSION THREAD

**Discussion by Leopold95** [DerFrZocker/Ore-Control#28] [Q&A] [answered]
Hello! I'm trying to create my own ore control plugin. I've analyzed your source a lot, but I can't get into it to understand the base algorithm. Can I get help from you? This is like explaining the basics of the ore control you use. I will be very grateful!

**Comment by DerFrZocker** [comment, marked as answer]
Minecraft has for every ore it generates one or more different PlacedFeature's which holds the information what ore should be generated and where and how often it should generate.

The where part is thereby calculated with multiple PlacementModifier's (PlacedFeature contains a list of those), during generation minecraft picks a block from the chunk it generates and passes it through every PlacementModifier present in the PlacedFeature. At the end it has a list of positions in the chunk where it generates the ore.

Example list of some PlacementModifier's:
- CountPlacement -> Is used to configured how many ores should generated in a chunk
- HeightRangePlacement -> Is used to spreed the y positions between two specific values (with the option of different distribution)
- InSquarePlacement -> Distributes the x and z position over the whole 16x16 chunk
- BiomeFilter -> Controls that ores are only generated in the biomes they should (Can be used to generated different amounts of ores based on the biome)

For the actual ore vein generation the PlacedFeature contains a ConfiguredFeature, which itself contains a Feature and a FeatureConfiguration. The Feature is thereby OreFeature with OreConfiguration as FeatureConfiguration.

The OreConfiguration contains options to set the size of the vein, as well as the discard chance on air exposure (see [here](https://github.com/DerFrZocker/Ore-Control/wiki/Generator:-Minecraft:-Ore#pictures) for more infos). It also contains a list of TargetBlockState which control which block should be replaced with which ore. For example, generated normal ore when replacing stone, but use the deepslate variant of the ore when replacing deepslate.

To actually change the generation you now first create your own PlacedFeature with the parameters you want.

Now you need to add your custom PlacedFeature to the biomes generation setting (which you can find the code of [here](https://github.com/DerFrZocker/Ore-Control/blob/b4e889e9c9751ac00a6431e3efa2dde1517a7188/impl/v1_20_R3/src/main/java/de/derfrzocker/ore/control/impl/v1_20_R3/NMSReplacer_v1_20_R3.java#L418)) and also remove the ore you don't want to to have generated in the biome. For this you need to create a new List of HolderSet<PlacedFeature> and set it to BiomeGenerationSettings#features and BiomeGenerationSettings#featureSet, but don't forget to also add the none ores features of the original BiomeGenerationSettings#features list and also make sure to preserver the current order of it.

With that you have modified the ore generation. Note however, that this method is very hacky and can mess up the server (although I didn't have any problems with it so far).

**Reply by Leopold95** [reply]
Thank you very much for this explanation. This is useful information for versions 1.17+. 
If you don't mind, could you explain a similar topic about how to control ores, for version 1.16? So, there are no similar 1.17 innovations. Sorry, if i ask stupid questions, but there no information in internet about at all.
Thank you in advance!
